commands/cleanupwatchdog.c

```diff
@@ -259,32 +263,35 @@ do_watchdog_hypercall(uint32_t *id, uint32_t timeout)
         return MTC_ERROR_WD_INSUFFICIENT_RESOURCE;
     }
     ret = ioctl(fd, IOCTL_PRIVCMD_HYPERCALL, &hypercall);
-    if (ret < 0) 
+
+    if (ret != 0)
     {
+        ret = ret < 0 ? errno : ret;
+
         close(fd);
         unlock_pages(&hypercall, sizeof(hypercall));
         unlock_pages(&arg, sizeof(arg));
-        return MTC_ERROR_WD_INSTANCE_UNAVAILABLE ;
-    }
-    close(fd);
-    unlock_pages(&hypercall, sizeof(hypercall));
-    unlock_pages(&arg, sizeof(arg));
-
-    //
-    // if id == 0 ret is new id should be > 0
-    //
 
-    if (*id == 0) 
-    {
-        if (ret > 0) 
+        if (ret == -EINVAL)
         {
-            *id = ret;
+            // This may happen because the slot is not in use
+            // or because there is no slot with this id
+            return MTC_ERROR_WD_INVALID_HANDLE;
         }
-        else 
+        if (ret > 0)
         {
-            return MTC_ERROR_WD_INSTANCE_UNAVAILABLE;
+            // A new watchdog was set, this is an
+            // This should not be possible since
+            // *id is made sure to not be 0
+            return MTC_ERROR_UNDEFINED;
         }
+        return MTC_ERROR_WD_INSTANCE_UNAVAILABLE;
     }
+
+    close(fd);
```

Comment:
Would error handling be clearer by using `goto out` and assigning to `rc` to in the code that checks the result of calls?

```
out:
  close(fd);
  unlock_pages(&hypercall, sizeof(hypercall));
  unlock_pages(&arg, sizeof(arg));
  return rc;
```

Reply:
I can remove the repetition with an `rc` variable, I don't think the label is needed seeing that flow in this function is quite simple.

Reply:
I would prefer a still flatter structure:
```
  ret = ioctl(fd, IOCTL_PRIVCMD_HYPERCALL, &hypercall);

  if (ret == 0)
    rc = MTC_SUCCESS;
  else if (ret < 0 && errno == -EINVAL)
    rc = MTC_ERROR_WD_INVALID_HANDLE;
  else if (ret < 0)
    rc = MTC_ERROR_WD_INSTANCE_UNAVAILABLE;
  else  (* ret > 0 *)
    rc = MTC_ERROR_UNDEFINED;
    
  close(fd);
  unlock_pages(&hypercall, sizeof(hypercall));
  unlock_pages(&arg, sizeof(arg));

  return rc;  
```